### PR TITLE
Features/584 data parallel -- DASO

### PR DESCRIPTION
**Pull request by coquelin77**
## Description

Implementation of the Distributed Asynchronous and Selective Optimization (`DASO`) method. 
Other changes include a second Image-Net example using `DALI` on `TFRecords` with `DASO`, `MPI` sum functions for `torch.half` and `torch.bfloat16`, as well as a class to determine if a metric is plateauing.

Issue/s resolved: N/A

## Changes proposed:
- DASO implementation
- MPI sum functions for `torch.half` and `torch.bfloat16` dtypes
- `nn.DataParallelMultiGPU`: utilize `torch.distributed` on a node-local level for the forward and backward steps of a NN
- `optim.DetectMetricPlateau` can be used to detect when a metric has plateaued. this is often used during training of the optimizers

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by coquelin77** [comment]
rerun tests

**Comment by coquelin77** [comment]
rerun tests

File: heat/optim/dp_optimizer.py [hunk 1]
```diff
@@ -1,6 +1,788 @@
...
+    new_buff = MPI.memory.fromaddress(tens_b.data_ptr(), nbytes=nelem * tens_b.element_size())
+    buffer_b[:] = new_buff
+
+
+# create new MPI OPs
+mpi_sum_f16 = MPI.Op.Create(__sum_f16_cb, commute=True)
+mpi_sum_bfloat = MPI.Op.Create(__sum_bfloat_cb, commute=True)
+
+
+class DASO:
```
Comment (Markus-Goetz) [inline heat/optim/dp_optimizer.py:42]: Type annotations sometimes available, sometimes not. Is there a reason for this?

File: heat/optim/dp_optimizer.py [hunk 2]
```diff
@@ -1,6 +1,788 @@
...
+
+        Parameters
+        ----------
+        model: torch.nn.Module
+            the local torch model.
+        """
+        self.module = model
+
+    def _start_local_sync(self):
+        # *start* local synchronizations for the next batches
```
Comment (Markus-Goetz) [inline heat/optim/dp_optimizer.py:680]: inconsistent doc strings

**Review by Markus-Goetz — CHANGES_REQUESTED** [review]

File: heat/nn/data_parallel.py [hunk 3]
```diff
@@ -70,6 +80,10 @@ def __init__(
         optimizer: Union[optim.DataParallelOptimizer, List, Tuple],
         blocking_parameter_updates: bool = False,
     ):
+        if isinstance(optimizer, optim.DASO):
+            raise TypeError(
+                "For use with DASO please use DataParallelMultiGPU instead of DataParallel"
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/nn/data_parallel.py:85]: i think that would take too much work to unify them. they use vastly different methods of communication. So i think its easier to just throw the error. and its also at the top of the code so that helps

File: heat/nn/data_parallel.py [hunk 4]
```diff
@@ -309,3 +330,68 @@ def _reset_parameters(module: tnn.Module) -> None:
...
+        module: torch.nn.Module
+            an implemented PyTorch model
+        optimizer: optim.DASO
+            A DASO optimizer. Other optimizers are not yet implemented. The DASO optimizer should be
+            defined prior to calling this class.
+        comm: MPICommunication, optional
+            A global communicator.
+            Default: ht.MPICommunication
+        """
+        torch.random.manual_seed(2147483646)  # max int32 value - 1
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/nn/data_parallel.py:358]: the idea was to unify the model on each process during the init. but i guess that isnt always needed and can be done when defining the local model before this

File: examples/nn/imagenet-DASO.py [hunk 5]
```diff
@@ -0,0 +1,864 @@
...
+    import nvidia.dali as dali
+    import nvidia.dali.ops as ops
+    import nvidia.dali.tfrecord as tfrec
+except ImportError:
+    raise ImportError(
+        "Please install DALI from https://www.github.com/NVIDIA/DALI to run this example."
+    )
+
+
+def print0(*args, **kwargs):
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply examples/nn/imagenet-DASO.py:36]: maybe in like a `debug.py` file. that could make it a bit easier for us to print things

[the remark this answers is not in the document]
Comment (coquelin77) [reply examples/nn/imagenet-DASO.py:36]: ill leave it here for now and make a separate issue for it

File: examples/nn/imagenet-DASO.py [hunk 6]
```diff
@@ -0,0 +1,864 @@
+import argparse
+import os
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply examples/nn/imagenet-DASO.py:2]: agreed. an NN example is on the long list of things that i need to do still

File: heat/optim/dp_optimizer.py (hunk 1 above)
Comment (coquelin77) [reply heat/optim/dp_optimizer.py:42]: nothing aside from documenting at different times. will correct now

File: heat/core/communication.py [hunk 7]
```diff
@@ -59,16 +59,28 @@ class MPICommunication(Communication):
...
         torch.int32: MPI.INT,
         torch.int64: MPI.LONG,
+        torch.bfloat16: MPI.SHORT,
+        torch.float16: MPI.SHORT,
         torch.float32: MPI.FLOAT,
         torch.float64: MPI.DOUBLE,
         torch.complex64: MPI.COMPLEX,
         torch.complex128: MPI.DOUBLE_COMPLEX,
     }
+    MPI._typedict["e"] = __mpi_type_mappings
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/communication.py]: relic of other attempts. when removed, everything still runs normally.

File: heat/optim/dp_optimizer.py [hunk 8]
```diff
@@ -1,6 +1,788 @@
...
+    tens_a = torch.HalfTensor().set_(torch.HalfStorage.from_buffer(buffer_a, "native"))
+    tens_b = torch.HalfTensor().set_(torch.HalfStorage.from_buffer(buffer_b, "native"))
+    tens_b += tens_a
+    nelem = torch.prod(torch.tensor(tens_b.shape)).item()
+    new_buff = MPI.memory.fromaddress(tens_b.data_ptr(), nbytes=tens_b.element_size() * nelem)
+    buffer_b[:] = new_buff
+
+
+def __sum_bfloat_cb(buffer_a, buffer_b, _):
+    # MPI custom sum function to use torch.bfloat16
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/optim/dp_optimizer.py:28]: im not sure that i can do that with the syntax for MPI custom ops

File: heat/optim/dp_optimizer.py [hunk 9]
```diff
@@ -1,6 +1,788 @@
...
+
+        self.cast_dtype = downcast_type
+        if downcast_type == torch.bfloat16:
+            self.cast_fn = mpi_sum_bfloat
+        elif downcast_type == torch.half:
+            self.cast_fn = mpi_sum_f16
+        else:
+            self.cast_fn = MPI.SUM
+
+        self.comm = MPI_WORLD
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/optim/dp_optimizer.py:156]: why? im not sure that this will help anything. i think it might cause issues if partial comms are used because DASO depends on homogeneous nodes

File: heat/optim/dp_optimizer.py [hunk 10]
```diff
@@ -1,6 +1,788 @@
...
+
+        self.print0(
+            f"Best loss value: {self.stability.best * (1.0 - self.stability.threshold):.4f}"
+            f" Current loss: {avg_loss:.4f}, Worse epochs: {self.stability.num_bad_epochs}"
+        )
+
+        stable = self.stability.test_if_improving(avg_loss)
+
+        if stable and self.global_skip > 1:
+            # drop gs by factor of 2
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/optim/dp_optimizer.py:369]: im confused by this comment. what do you mean?

File: heat/core/communication.py [hunk 11]
```diff
@@ -59,16 +59,28 @@ class MPICommunication(Communication):
...
 
-    def __init__(self, handle=MPI.COMM_WORLD):
+    def __init__(self, handle=MPI.COMM_WORLD, group=False):
         self.handle = handle
-        self.rank = handle.Get_rank()
-        self.size = handle.Get_size()
+        if group:
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/core/communication.py]: updated to 
```python
    def __init__(self, handle=MPI.COMM_WORLD):
        self.handle = handle
        try:
            self.rank = handle.Get_rank()
            self.size = handle.Get_size()
        except MPI.Exception:
            # ranks not within the group will fail with an MPI.Exception, this is expected
            self.rank = None
            self.size = None
```

File: examples/nn/imagenet-DASO.py [hunk 12]
```diff
@@ -0,0 +1,864 @@
...
+try:
+    from nvidia.dali.plugin.pytorch import DALIClassificationIterator
+    from nvidia.dali.pipeline import Pipeline
+    import nvidia.dali as dali
+    import nvidia.dali.ops as ops
+    import nvidia.dali.tfrecord as tfrec
+except ImportError:
+    raise ImportError(
+        "Please install DALI from https://www.github.com/NVIDIA/DALI to run this example."
+    )
```
[the remark this answers is not in the document]
Comment (coquelin77) [reply examples/nn/imagenet-DASO.py:34]: i added a finalize before the raise

File: examples/nn/imagenet-DASO.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (coquelin77) [reply examples/nn/imagenet-DASO.py:37]: see #739

File: examples/nn/imagenet-DASO.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply examples/nn/imagenet-DASO.py:2]: Perfect, issue already opened. Nice!

File: examples/nn/imagenet-DASO.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply examples/nn/imagenet-DASO.py:34]: Better, but still not quite where I would want to have it. I thought more like along the lines of:

```
try:
    ...
except ImportError:
    import sys
    sys.stderr.write("Please install...")  # perhaps a print0 to sys.stderr instead
    ht.MPI.Finalize()
    sys.exit(0)
```

This way we will prevent the spamy MPI output for a process having non-zero exit code.

File: examples/nn/imagenet-DASO.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply examples/nn/imagenet-DASO.py:37]: perfect

File: heat/nn/data_parallel.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/nn/data_parallel.py:86]: agreed

File: heat/optim/dp_optimizer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/optim/dp_optimizer.py:28]: How about

```
def __sum_f16_general(tensor_cls, buffer_a, buffer_b):
    tens_a = tensor_cls().set_(....)
    # remaining code

def __sum_f16_cb(buffer_a, buffer_b):
    return __sum_f16_general(HalfTensor, buffer_a, buffer_b)
```

File: heat/optim/dp_optimizer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/optim/dp_optimizer.py:156]: Because all the other heat interfaces can do that. Perhaps a user warning and/or a remark in the docstring about this?

File: heat/optim/dp_optimizer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (Markus-Goetz) [reply heat/optim/dp_optimizer.py:371]: The factor 2 should be a global drop constant at the top of the file or class constant so that it can be easily changed in the future. Same goes for the other constants below, like the // 4

File: heat/optim/dp_optimizer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (coquelin77) [reply heat/optim/dp_optimizer.py:28]: this introduces NaNs for unknown reasons.

**Review by Markus-Goetz — APPROVED** [review]
